### ISSUE TITLE
Suggest replacing simple loops with specialized folds

### ISSUE THREAD

**Issue by killercup** [rust-lang-nursery/rust-clippy#2401]
A lot of simple loops can be written using fold-like methods in iterators. Clippy should have a set of lints (maybe even a lint group[^1]) that promotes the more functional style.

Have a look at this code:

```rust
let xs = &[1, 2, 3, 4]; // <- ignore that this is a slice for a second

let mut res = vec![];
for x in xs {
    res.push(x);
}
```

Seasoned functional programmers recognize that is is just a fold/reduce. In Rust, you'd write it like this:

```rust
xs.iter().fold(vec![], |mut acc, x| { acc.push(x); acc })
```

Seasoned _Rust_ programmers would use a specialized fold here (basically a shortcut that makes use of FromIterator):

```rust
let res: Vec<_> = xs.iter().collect();
```

- [ ] clippy should suggest using `.collect::<Vec<_>>()` for loops that just build vectors

Let's make this a bit more complicated:

```rust
let xs = &[1, 2, 3, 4];

let mut res = vec![];
for x in xs {
    if *x > 2 { res.push(x); } 
}
```

This is still a pretty simple loop, and we can rewrite it like this:

```rust
let res: Vec<_> = xs.iter().filter(|&x| *x > 2).collect();
```

(Sadly, it's not a trivial as it first appeared to be because filter passes a reference to the closure, and we need to rewrite that in the comparison we copied from the `if`. Instead of `|&x| *x > 2` we could also generate `|x| **x > 2` or `|&&x| x > 2`.)

- [ ] clippy should suggest using `.filter.collect` for these kinds of loops

There are of course a bunch more specialized folders we can recognize. For example:

```rust
let mut res = false;
for x in xs {
    if *x > 2 { res = true; } 
}
```

Isn't it wonderfully procedural? It takes 4 lines and inline mutation to find out if an item in that iterator is greater than two. How about suggesting `any` instead? (Also see [this exisiting lint](https://rust-lang-nursery.github.io/rust-clippy/v0.0.181/index.html#unnecessary_fold).)

```rust
let res = xs.iter().any(|x| *x > 2);
```

- [ ] clippy should suggest using `any` for these loops

---

Haven't found an open issue about that, searched for "loop", "reduce", "fold".

[^1]: Please name the group after an [Office Assistant](https://en.wikipedia.org/wiki/Office_Assistant#Assistants)!

**Comment by shaleh** [comment]
This is a great idea. Whenever I bring up `fold` it is like I unveiled an unknown maths or something :-)

**Comment by scottmcm** [comment]
And as a bonus, `fold` (or `all` or `position` or ...) is sometimes faster than `for`, and is never slower.

**Comment by TheIronBorn** [comment]
> Instead of `|&x| *x > 2` we could also generate `|x| **x > 2` or `|&&x| x > 2`

I prefer `|&&x|` in most situations because I find it removes the need to type `*x` for each mention (similar to [match_ref_pats](https://rust-lang-nursery.github.io/rust-clippy/v0.0.188/index.html#match_ref_pats)).  It doesn't work for every case though.
